<a href="https://colab.research.google.com/github/chanxxw/AIFFEL_quest_cr/blob/master/Python/Project06.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Project 01

In [31]:
import sklearn
print(sklearn.__version__)

1.3.2


In [32]:
from sklearn.datasets import load_digits
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report
import pandas as pd
import numpy as np

In [33]:
type(load_digits) # 타입 확인

function

In [34]:
digits=load_digits() # digit 내용을 불러와서 저장하기

In [35]:
digits.keys() # dataset 정보 종류 확인

dict_keys(['data', 'target', 'frame', 'feature_names', 'target_names', 'images', 'DESCR'])

In [36]:
digits.data # digits의 실제 데이터 확인

array([[ 0.,  0.,  5., ...,  0.,  0.,  0.],
       [ 0.,  0.,  0., ..., 10.,  0.,  0.],
       [ 0.,  0.,  0., ..., 16.,  9.,  0.],
       ...,
       [ 0.,  0.,  1., ...,  6.,  0.,  0.],
       [ 0.,  0.,  2., ..., 12.,  0.,  0.],
       [ 0.,  0., 10., ..., 12.,  1.,  0.]])

In [37]:
digits_data=digits.data # digits의 data를 digits_data에 저장
digits_label = digits.target # key에서 확인한 정보중 target을 digits_label 변수에 저장

In [39]:
digits_label # digit의 subgroup 종류 (=digit.target)

array([0, 1, 2, ..., 8, 9, 8])

In [40]:
digits.feature_names # digit data의 이름

['pixel_0_0',
 'pixel_0_1',
 'pixel_0_2',
 'pixel_0_3',
 'pixel_0_4',
 'pixel_0_5',
 'pixel_0_6',
 'pixel_0_7',
 'pixel_1_0',
 'pixel_1_1',
 'pixel_1_2',
 'pixel_1_3',
 'pixel_1_4',
 'pixel_1_5',
 'pixel_1_6',
 'pixel_1_7',
 'pixel_2_0',
 'pixel_2_1',
 'pixel_2_2',
 'pixel_2_3',
 'pixel_2_4',
 'pixel_2_5',
 'pixel_2_6',
 'pixel_2_7',
 'pixel_3_0',
 'pixel_3_1',
 'pixel_3_2',
 'pixel_3_3',
 'pixel_3_4',
 'pixel_3_5',
 'pixel_3_6',
 'pixel_3_7',
 'pixel_4_0',
 'pixel_4_1',
 'pixel_4_2',
 'pixel_4_3',
 'pixel_4_4',
 'pixel_4_5',
 'pixel_4_6',
 'pixel_4_7',
 'pixel_5_0',
 'pixel_5_1',
 'pixel_5_2',
 'pixel_5_3',
 'pixel_5_4',
 'pixel_5_5',
 'pixel_5_6',
 'pixel_5_7',
 'pixel_6_0',
 'pixel_6_1',
 'pixel_6_2',
 'pixel_6_3',
 'pixel_6_4',
 'pixel_6_5',
 'pixel_6_6',
 'pixel_6_7',
 'pixel_7_0',
 'pixel_7_1',
 'pixel_7_2',
 'pixel_7_3',
 'pixel_7_4',
 'pixel_7_5',
 'pixel_7_6',
 'pixel_7_7']

In [41]:
digits.filename # 오류군 왜지??

AttributeError: filename

In [42]:
digits_df = pd.DataFrame(data=digits.data, columns=digits.feature_names) #데이터프레임 소환, 이것땜에 위에 판다스 추가 import
digits_df['label'] = digits.target # 라벨 추가(정답지)
digits_df.head() # 상위 5개만 확인해보기

pixel_0_0  pixel_0_1  pixel_0_2  pixel_0_3  pixel_0_4  pixel_0_5  \
0        0.0        0.0        5.0       13.0        9.0        1.0   
1        0.0        0.0        0.0       12.0       13.0        5.0   
2        0.0        0.0        0.0        4.0       15.0       12.0   
3        0.0        0.0        7.0       15.0       13.0        1.0   
4        0.0        0.0        0.0        1.0       11.0        0.0   

   pixel_0_6  pixel_0_7  pixel_1_0  pixel_1_1  ...  pixel_6_7  pixel_7_0  \
0        0.0        0.0        0.0        0.0  ...        0.0        0.0   
1        0.0        0.0        0.0        0.0  ...        0.0        0.0   
2        0.0        0.0        0.0        0.0  ...        0.0        0.0   
3        0.0        0.0        0.0        8.0  ...        0.0        0.0   
4        0.0        0.0        0.0        0.0  ...        0.0        0.0   

   pixel_7_1  pixel_7_2  pixel_7_3  pixel_7_4  pixel_7_5  pixel_7_6  \
0        0.0        6.0       13.0       10.0        0.0        0.0   
1        0.0        0.0       11.0       16.0       10.0        0.0   
2        0.0        0.0        3.0       11.0       16.0        9.0   
3        0.0        7.0       13.0       13.0        9.0        0.0   
4        0.0        0.0        2.0       16.0        4.0        0.0   

   pixel_7_7  label  
0        0.0      0  
1        0.0      1  
2        0.0      2  
3        0.0      3  
4        0.0      4  

[5 rows x 65 columns]

In [43]:
digits_df.keys()

Index(['pixel_0_0', 'pixel_0_1', 'pixel_0_2', 'pixel_0_3', 'pixel_0_4',
       'pixel_0_5', 'pixel_0_6', 'pixel_0_7', 'pixel_1_0', 'pixel_1_1',
       'pixel_1_2', 'pixel_1_3', 'pixel_1_4', 'pixel_1_5', 'pixel_1_6',
       'pixel_1_7', 'pixel_2_0', 'pixel_2_1', 'pixel_2_2', 'pixel_2_3',
       'pixel_2_4', 'pixel_2_5', 'pixel_2_6', 'pixel_2_7', 'pixel_3_0',
       'pixel_3_1', 'pixel_3_2', 'pixel_3_3', 'pixel_3_4', 'pixel_3_5',
       'pixel_3_6', 'pixel_3_7', 'pixel_4_0', 'pixel_4_1', 'pixel_4_2',
       'pixel_4_3', 'pixel_4_4', 'pixel_4_5', 'pixel_4_6', 'pixel_4_7',
       'pixel_5_0', 'pixel_5_1', 'pixel_5_2', 'pixel_5_3', 'pixel_5_4',
       'pixel_5_5', 'pixel_5_6', 'pixel_5_7', 'pixel_6_0', 'pixel_6_1',
       'pixel_6_2', 'pixel_6_3', 'pixel_6_4', 'pixel_6_5', 'pixel_6_6',
       'pixel_6_7', 'pixel_7_0', 'pixel_7_1', 'pixel_7_2', 'pixel_7_3',
       'pixel_7_4', 'pixel_7_5', 'pixel_7_6', 'pixel_7_7', 'label'],
      dtype='object')

In [44]:
digits_df.describe()

pixel_0_0    pixel_0_1    pixel_0_2    pixel_0_3    pixel_0_4  \
count     1797.0  1797.000000  1797.000000  1797.000000  1797.000000   
mean         0.0     0.303840     5.204786    11.835838    11.848080   
std          0.0     0.907192     4.754826     4.248842     4.287388   
min          0.0     0.000000     0.000000     0.000000     0.000000   
25%          0.0     0.000000     1.000000    10.000000    10.000000   
50%          0.0     0.000000     4.000000    13.000000    13.000000   
75%          0.0     0.000000     9.000000    15.000000    15.000000   
max          0.0     8.000000    16.000000    16.000000    16.000000   

         pixel_0_5    pixel_0_6    pixel_0_7    pixel_1_0    pixel_1_1  ...  \
count  1797.000000  1797.000000  1797.000000  1797.000000  1797.000000  ...   
mean      5.781859     1.362270     0.129661     0.005565     1.993879  ...   
std       5.666418     3.325775     1.037383     0.094222     3.196160  ...   
min       0.000000     0.000000     0.000000     0.000000     0.000000  ...   
25%       0.000000     0.000000     0.000000     0.000000     0.000000  ...   
50%       4.000000     0.000000     0.000000     0.000000     0.000000  ...   
75%      11.000000     0.000000     0.000000     0.000000     3.000000  ...   
max      16.000000    16.000000    15.000000     2.000000    16.000000  ...   

         pixel_6_7    pixel_7_0    pixel_7_1    pixel_7_2    pixel_7_3  \
count  1797.000000  1797.000000  1797.000000  1797.000000  1797.000000   
mean      0.206455     0.000556     0.279354     5.557596    12.089037   
std       0.984401     0.023590     0.934302     5.103019     4.374694   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.000000     0.000000     0.000000     1.000000    11.000000   
50%       0.000000     0.000000     0.000000     4.000000    13.000000   
75%       0.000000     0.000000     0.000000    10.000000    16.000000   
max      13.000000     1.000000     9.000000    16.000000    16.000000   

         pixel_7_4    pixel_7_5    pixel_7_6    pixel_7_7        label  
count  1797.000000  1797.000000  1797.000000  1797.000000  1797.000000  
mean     11.809126     6.764051     2.067891     0.364496     4.490818  
std       4.933947     5.900623     4.090548     1.860122     2.865304  
min       0.000000     0.000000     0.000000     0.000000     0.000000  
25%      10.000000     0.000000     0.000000     0.000000     2.000000  
50%      14.000000     6.000000     0.000000     0.000000     4.000000  
75%      16.000000    12.000000     2.000000     0.000000     7.000000  
max      16.000000    16.000000    16.000000    16.000000     9.000000  

[8 rows x 65 columns]

In [45]:
from sklearn.model_selection import train_test_split # train/test model 으로 dataset 구분
X_train, X_test, y_train, y_test=train_test_split (digits_data, digits_label, test_size = 0.2, random_state = 3)
# 나눠야 할 데이터(문제지, X): digits_data
# 데이터의 라벨(정답, y): digits_label
# digits_data와 digits_label를 각각 train:test = 8:2의 비율로 잘라서
# random state = 인수 지정한 것. 노드에선 2로 되어있으나 그냥 3으로 해봄
# X_train, X_test, y_train, y_test에 저장
print('X_train 개수: ', len(X_train),', X_test 개수: ', len(X_test)) # train/test 개수 확인

X_train 개수:  1437 , X_test 개수:  360


In [48]:
# Decision Tree 모델 import 하기
from sklearn.tree import DecisionTreeClassifier
decision_tree=DecisionTreeClassifier(random_state=10) #재현성을 위해 초기값 10, 바꿔봄
decision_tree.fit(X_train, y_train) # fit은 의사결정나무로 학습하기위한 명령어
y_pred = decision_tree.predict(X_test)
y_pred


array([0, 4, 1, 2, 0, 0, 8, 7, 6, 6, 3, 6, 9, 7, 4, 7, 4, 3, 2, 6, 7, 3,
       4, 7, 1, 0, 7, 4, 8, 3, 4, 0, 5, 5, 5, 1, 2, 9, 0, 0, 0, 8, 2, 3,
       7, 0, 2, 7, 1, 3, 8, 4, 2, 9, 6, 0, 4, 3, 7, 4, 3, 9, 9, 5, 0, 3,
       7, 4, 9, 1, 3, 0, 9, 3, 0, 2, 7, 8, 1, 4, 1, 3, 3, 2, 2, 3, 8, 2,
       4, 6, 9, 5, 8, 4, 7, 2, 7, 2, 5, 6, 5, 6, 1, 0, 6, 4, 3, 1, 5, 6,
       8, 8, 8, 6, 0, 0, 2, 7, 9, 0, 8, 5, 8, 4, 2, 0, 5, 9, 0, 1, 8, 4,
       4, 3, 6, 2, 2, 7, 1, 3, 5, 8, 9, 2, 7, 4, 9, 2, 0, 0, 9, 2, 3, 4,
       0, 9, 7, 0, 1, 4, 1, 3, 6, 7, 9, 1, 9, 7, 2, 7, 7, 0, 5, 3, 4, 0,
       5, 3, 8, 0, 3, 0, 1, 9, 5, 4, 8, 2, 6, 0, 9, 7, 8, 8, 7, 3, 0, 3,
       9, 3, 6, 3, 7, 9, 0, 2, 7, 6, 5, 3, 1, 0, 4, 7, 1, 0, 5, 3, 7, 3,
       2, 3, 5, 0, 7, 4, 5, 3, 0, 0, 5, 7, 5, 2, 4, 2, 9, 3, 3, 0, 9, 3,
       2, 3, 1, 7, 0, 7, 4, 5, 2, 1, 1, 4, 7, 2, 7, 8, 3, 5, 2, 3, 0, 9,
       2, 0, 7, 6, 5, 4, 3, 1, 4, 1, 8, 0, 7, 4, 2, 1, 2, 4, 9, 5, 0, 3,
       0, 5, 5, 3, 6, 1, 4, 3, 8, 5, 7, 7, 4, 6, 4,

In [49]:
y_test

array([0, 4, 1, 2, 0, 0, 8, 7, 6, 6, 3, 6, 9, 7, 4, 7, 4, 3, 2, 6, 4, 3,
       4, 7, 1, 0, 7, 4, 8, 3, 4, 0, 5, 5, 5, 1, 2, 9, 0, 0, 0, 8, 2, 3,
       7, 0, 1, 7, 1, 3, 8, 4, 2, 9, 6, 0, 4, 5, 4, 4, 7, 9, 9, 5, 0, 3,
       7, 4, 9, 1, 8, 0, 9, 3, 0, 2, 7, 8, 1, 6, 9, 3, 3, 2, 2, 3, 8, 2,
       4, 6, 9, 5, 8, 4, 7, 3, 7, 2, 5, 6, 5, 6, 1, 0, 6, 4, 8, 1, 5, 6,
       8, 8, 8, 6, 4, 0, 2, 7, 5, 0, 8, 5, 8, 4, 7, 0, 5, 9, 0, 1, 8, 4,
       7, 9, 6, 1, 2, 7, 1, 3, 5, 3, 5, 2, 7, 4, 9, 2, 0, 0, 9, 2, 8, 4,
       0, 9, 7, 0, 1, 4, 1, 8, 0, 7, 9, 1, 9, 7, 2, 7, 7, 0, 5, 3, 4, 0,
       5, 2, 3, 0, 3, 0, 1, 9, 5, 4, 8, 2, 6, 0, 9, 7, 8, 8, 7, 3, 0, 9,
       5, 3, 6, 3, 7, 9, 0, 5, 7, 6, 5, 3, 1, 0, 4, 3, 1, 0, 5, 3, 7, 3,
       2, 3, 5, 0, 7, 4, 5, 3, 0, 0, 5, 7, 5, 2, 4, 2, 9, 3, 3, 0, 9, 3,
       1, 3, 1, 7, 4, 7, 4, 5, 2, 1, 1, 4, 7, 1, 7, 5, 2, 5, 2, 5, 0, 9,
       2, 0, 7, 6, 5, 4, 9, 1, 4, 1, 8, 0, 7, 4, 7, 1, 3, 4, 9, 5, 0, 3,
       0, 5, 4, 3, 6, 1, 4, 3, 8, 5, 2, 7, 4, 6, 4,

In [50]:
# Accuracy 평가 위한 model
from sklearn.metrics import accuracy_score
accuracy = accuracy_score(y_test, y_pred)
accuracy

0.8611111111111112

In [52]:
from sklearn.ensemble import RandomForestClassifier #랜덤포레스트라는 분류기를 사용하기 위해 import

X_train, X_test, y_train, y_test = train_test_split(digits_data, # digits 데이터의 data 컬럼
                                                    digits_label, # digits 데이터의 target 컬럼
                                                    test_size=0.2, # test_size : train data와 test data를 몇대몇으로 나눌지 정하는 옵션
                                                    random_state=21) # random_state : 랜덤 패턴의 값을 지정

random_forest = RandomForestClassifier(random_state=32) # RandomForest분류기 객체를 생성
random_forest.fit(X_train, y_train) # 훈련
y_pred = random_forest.predict(X_test) # 예측

print(classification_report(y_test, y_pred)) # 결과 지표를 확인

              precision    recall  f1-score   support

           0       0.97      0.94      0.95        32
           1       0.97      1.00      0.99        36
           2       1.00      1.00      1.00        30
           3       0.98      0.98      0.98        41
           4       0.94      0.97      0.95        32
           5       1.00      0.98      0.99        46
           6       1.00      0.97      0.98        32
           7       0.98      1.00      0.99        40
           8       0.93      0.98      0.95        42
           9       1.00      0.93      0.96        29

    accuracy                           0.97       360
   macro avg       0.98      0.97      0.97       360
weighted avg       0.98      0.97      0.98       360



In [53]:
from sklearn import svm #Support Vector Machine을 사용하기 위해 import
svm_model = svm.SVC() # 모델 객체를 만든다.

print(svm_model._estimator_type) # 이 모델이 어떤 타입인지 확인한다.

svm_model.fit(X_train, y_train) # 훈련
y_pred = svm_model.predict(X_test) # 예측

print(classification_report(y_test, y_pred)) # 결과 지표를 확인


classifier
              precision    recall  f1-score   support

           0       1.00      0.97      0.98        32
           1       0.95      1.00      0.97        36
           2       1.00      1.00      1.00        30
           3       1.00      1.00      1.00        41
           4       0.97      1.00      0.98        32
           5       0.98      1.00      0.99        46
           6       1.00      1.00      1.00        32
           7       1.00      1.00      1.00        40
           8       0.98      0.95      0.96        42
           9       1.00      0.93      0.96        29

    accuracy                           0.99       360
   macro avg       0.99      0.99      0.99       360
weighted avg       0.99      0.99      0.99       360



In [54]:
from sklearn.linear_model import SGDClassifier #선형분류기인 SGDClassifier를 사용하기 위한 import
sgd_model = SGDClassifier() # 모델 객체 생성

print(sgd_model._estimator_type) # 이 모델의 타입을 확인
sgd_model.fit(X_train, y_train) # sgd모델로 훈련데이터로 훈련시킨다.
y_pred = sgd_model.predict(X_test)# 그 모델로 test데이터를 사용해 예측

print(classification_report(y_test, y_pred)) # 결과 지표를 확인

classifier
              precision    recall  f1-score   support

           0       1.00      0.94      0.97        32
           1       0.97      0.94      0.96        36
           2       1.00      0.97      0.98        30
           3       0.95      0.98      0.96        41
           4       0.94      0.97      0.95        32
           5       0.98      0.98      0.98        46
           6       1.00      1.00      1.00        32
           7       0.97      0.95      0.96        40
           8       0.91      0.98      0.94        42
           9       0.93      0.93      0.93        29

    accuracy                           0.96       360
   macro avg       0.97      0.96      0.96       360
weighted avg       0.96      0.96      0.96       360



In [55]:
from sklearn.linear_model import LogisticRegression # 선형분류기인 LogisticRegression를 사용하기 위한 import
logistic_model = LogisticRegression() # 모델 객체 생성

print(logistic_model._estimator_type) # 이 모델의 타입을 확인

logistic_model.fit(X_train, y_train) #LogisticRegression모델로 훈련데이터를 가지고 훈련시킨다.
y_pred = logistic_model.predict(X_test) # 예측

print(classification_report(y_test, y_pred)) # 결과 지표를 확인

classifier
              precision    recall  f1-score   support

           0       1.00      1.00      1.00        32
           1       0.95      0.97      0.96        36
           2       1.00      1.00      1.00        30
           3       0.98      1.00      0.99        41
           4       0.94      0.97      0.95        32
           5       0.98      0.98      0.98        46
           6       1.00      1.00      1.00        32
           7       0.97      0.97      0.97        40
           8       1.00      0.95      0.98        42
           9       0.96      0.93      0.95        29

    accuracy                           0.98       360
   macro avg       0.98      0.98      0.98       360
weighted avg       0.98      0.98      0.98       360



/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:460: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


## 모델을 평가해보기..  
- 사실 테스트데이터를 전혀 이해하지 못해서, 어떤 결과가 더 좋은건지 모르겠음

# GPT 답변  
- Scikit-learn에서 제공하는 **Digits 데이터셋**(`load_digits`)은 0부터 9까지의 손으로 쓴 숫자를 분류하는 데이터셋입니다. 이미지 분류 문제이며, 각 이미지는 8x8 픽셀로 표현됩니다. **클래스가 10개**로 구성되어 있어 다중 클래스 분류 문제입니다. Digits 데이터셋은 **크기가 상대적으로 작고**, **특징 수가 적당히 많은** 편이라 여러 가지 머신러닝 모델을 비교하기에 적합합니다.

다섯 가지 모델을 비교할 때, Digits 데이터셋에 적합한 이유를 설명하겠습니다.

### 1. **SGDClassifier (Stochastic Gradient Descent)**
   - **특징**: 확률적 경사 하강법(SGD)을 사용하여 선형 분류기를 학습합니다.
   - **장점**: 대규모 데이터셋에서 학습이 빠르며, 특히 희소 데이터에 적합합니다.
   - **단점**: Digits 데이터셋처럼 작은 크기의 데이터에서는 SGD가 최적의 성능을 내지 못할 수 있습니다.
   - **추천**: Digits 데이터셋은 크기가 작고 복잡성이 낮기 때문에, SGD보다는 다른 모델이 더 적합할 수 있습니다.

### 2. **SVM (Support Vector Machine)**
   - **특징**: SVM은 데이터가 선형적으로 분리되지 않는 경우에도 커널을 사용해 고차원 공간에서 분류할 수 있습니다.
   - **장점**: 작은 데이터셋에서 성능이 매우 좋으며, 특히 데이터가 선형적으로 분리되지 않을 때 뛰어난 성능을 발휘합니다.
   - **단점**: 데이터셋이 크면 학습 속도가 느려질 수 있지만, Digits 데이터셋은 작아서 이 문제는 크지 않습니다.
   - **추천**: Digits 데이터셋에서는 **SVM**이 매우 좋은 선택입니다. 성능이 높고 안정적인 결과를 기대할 수 있습니다.

### 3. **RandomForestClassifier**
   - **특징**: 여러 개의 결정 트리를 학습하여 앙상블 기법을 통해 예측하는 모델입니다.
   - **장점**: 과적합 방지에 강하며, 데이터 전처리 없이도 안정적인 성능을 제공합니다. 분류 문제에서 매우 강력한 모델입니다.
   - **단점**: 트리의 수가 많아지면 학습 시간이 조금 늘어날 수 있지만, Digits 데이터셋은 크기가 작아서 문제가 되지 않습니다.
   - **추천**: Digits 데이터셋에서 높은 정확도를 제공하며, 일반적으로 **RandomForestClassifier**는 매우 좋은 성능을 발휘할 수 있습니다.

### 4. **DecisionTreeClassifier**
   - **특징**: 의사결정 규칙을 기반으로 트리를 형성하여 분류를 수행합니다.
   - **장점**: 학습이 빠르고 해석이 쉬우며, 소규모 데이터셋에 적합합니다.
   - **단점**: 과적합되기 쉬우며, 데이터가 복잡해지면 성능이 떨어질 수 있습니다.
   - **추천**: Digits 데이터셋에 빠르게 적용할 수 있지만, **과적합을 방지하기 위해 트리 깊이를 제한**하는 것이 좋습니다.

### 5. **LogisticRegression**
   - **특징**: 선형 분류 모델로, 클래스 간 경계를 선형으로 구분합니다.
   - **장점**: 간단하고 빠르며, 적당한 성능을 보입니다. 다중 클래스 문제에 대해 `one-vs-rest` 방식으로 적용할 수 있습니다.
   - **단점**: 데이터가 선형적으로 분리되지 않으면 성능이 떨어질 수 있습니다.
   - **추천**: Digits 데이터셋에서 **빠르고 간단하게 사용할 수 있는 모델**로 적합하지만, 비선형 패턴이 많으면 성능이 떨어질 수 있습니다.

### 비교 분석

- **데이터 크기**: Digits 데이터셋은 1797개의 샘플과 64개의 특징(8x8 이미지)을 가지고 있습니다. 작은 크기의 데이터셋이기 때문에 **학습 속도**는 대부분의 모델에서 큰 문제가 되지 않습니다.
- **모델 해석**: **Logistic Regression**과 **Decision Tree**는 해석이 쉬운 모델입니다.
- **최고의 성능**: **SVM**과 **RandomForestClassifier**가 가장 높은 성능을 발휘할 가능성이 큽니다. 특히 SVM은 작은 데이터셋에서 매우 높은 성능을 보일 수 있습니다.
- **복잡도와 시간**: **RandomForestClassifier**는 상대적으로 복잡하고 시간이 조금 더 걸리지만, 높은 성능을 보장합니다. 반면, **Decision Tree**는 빠르지만 과적합 문제가 있을 수 있습니다.

### 결론

Digits 데이터셋에서 **최고의 성능**을 기대하려면:
- **SVM**이나 **RandomForestClassifier**가 가장 효율적인 선택입니다.

모델 해석의 용이성도 중요하다면:
- **Logistic Regression**이나 **Decision Tree**도 적합합니다.

성능을 최우선으로 한다면 **SVM**이 **Digits 데이터셋**에서는 매우 강력한 성능을 발휘할 수 있습니다. **RandomForestClassifier**도 안정적이고 높은 성능을 제공하므로 추천됩니다.

# Project 02

In [56]:
# (1) 필요한 모듈 import
from sklearn.datasets import load_wine #싸이킷런 데이터셋에 있는 wine를 불러오는 코드
from sklearn.model_selection import train_test_split #싸이킷런에 있는 데이터를 train과 test dataset으로 나누어주는 함수 import
from sklearn.tree import DecisionTreeClassifier # 싸이킷런에 있는 결정트리분류기를 사용하기 위해 불러오는 코드
from sklearn.metrics import classification_report #싸이킷런에 있는 분류 결과에 대한 시각화를 위해 쓰는 코드

# (2) 데이터 준비
wine = load_wine() #wine 데이터 전체를 불러온다.
wine_data = wine.data #wine데이터의 data컬럼을 분류해 wine_data 변수에 담는다.
wine_label = wine.target #wine데이터의 target컬럼을 분류해 wine_label 변수에 담는다.

# (3) train, test 데이터 분리
#train_test_split()를 사용하여 X값, y값을 각각 train data와 test data로 나눈다. 함수에 들어 갈 파라미터로는 x,y가 들어가고
#test_size는 몇대몇으로 나눌지 정하는 옵션, random_state는 랜덤 패턴의 값을 지정한다. (어떤 값을 넣어도 무방하다.)
X_train, X_test, y_train, y_test = train_test_split(wine_data,
                                                    wine_label,
                                                    test_size=0.2,
                                                    random_state=7)


# (4) 모델 학습 및 예측
decision_tree = DecisionTreeClassifier(random_state=32) #결정트리분류기의 객체를 만든다.
decision_tree.fit(X_train, y_train) # 분류기에 x와 y의 훈련 데이터를 넣어 훈련 시킨다.
y_pred = decision_tree.predict(X_test) # 훈련된 분류기에 X_test라는 테스트 데이터셋을 넣어 얼마나 예측했는지 확인한다.

print(classification_report(y_test, y_pred)) # 결과를 지표로 확인하기 위해 classification_report를 활용해 y_test, y_pred 값을 넣어 확인한다.

              precision    recall  f1-score   support

           0       1.00      1.00      1.00         7
           1       0.89      1.00      0.94        17
           2       1.00      0.83      0.91        12

    accuracy                           0.94        36
   macro avg       0.96      0.94      0.95        36
weighted avg       0.95      0.94      0.94        36



In [58]:
random_forest = RandomForestClassifier(random_state=2) # RandomForest분류기 객체를 생성
random_forest.fit(X_train, y_train) # 훈련
y_pred = random_forest.predict(X_test) # 예측

print(classification_report(y_test, y_pred)) # 결과 지표를 확인

              precision    recall  f1-score   support

           0       1.00      1.00      1.00         7
           1       1.00      1.00      1.00        17
           2       1.00      1.00      1.00        12

    accuracy                           1.00        36
   macro avg       1.00      1.00      1.00        36
weighted avg       1.00      1.00      1.00        36



In [59]:
from sklearn import svm #Support Vector Machine을 사용하기 위해 import
svm_model = svm.SVC() # 모델 객체를 만든다.

print(svm_model._estimator_type) # 이 모델이 어떤 타입인지 확인한다.

svm_model.fit(X_train, y_train) # 훈련
y_pred = svm_model.predict(X_test) # 예측

print(classification_report(y_test, y_pred)) # 결과 지표를 확인


classifier
              precision    recall  f1-score   support

           0       0.86      0.86      0.86         7
           1       0.58      0.88      0.70        17
           2       0.33      0.08      0.13        12

    accuracy                           0.61        36
   macro avg       0.59      0.61      0.56        36
weighted avg       0.55      0.61      0.54        36



In [60]:
from sklearn.linear_model import SGDClassifier #선형분류기인 SGDClassifier를 사용하기 위한 import
sgd_model = SGDClassifier() # 모델 객체 생성

print(sgd_model._estimator_type) # 이 모델의 타입을 확인
sgd_model.fit(X_train, y_train) # sgd모델로 훈련데이터로 훈련시킨다.
y_pred = sgd_model.predict(X_test)# 그 모델로 test데이터를 사용해 예측

print(classification_report(y_test, y_pred)) # 결과 지표를 확인

classifier
              precision    recall  f1-score   support

           0       0.88      1.00      0.93         7
           1       0.00      0.00      0.00        17
           2       0.43      1.00      0.60        12

    accuracy                           0.53        36
   macro avg       0.43      0.67      0.51        36
weighted avg       0.31      0.53      0.38        36



/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1471: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1471: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1471: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


In [61]:
from sklearn.linear_model import LogisticRegression # 선형분류기인 LogisticRegression를 사용하기 위한 import
logistic_model = LogisticRegression() # 모델 객체 생성

print(logistic_model._estimator_type) # 이 모델의 타입을 확인

logistic_model.fit(X_train, y_train) #LogisticRegression모델로 훈련데이터를 가지고 훈련시킨다.
y_pred = logistic_model.predict(X_test) # 예측

print(classification_report(y_test, y_pred)) # 결과 지표를 확인

classifier
              precision    recall  f1-score   support

           0       1.00      0.86      0.92         7
           1       0.94      1.00      0.97        17
           2       1.00      1.00      1.00        12

    accuracy                           0.97        36
   macro avg       0.98      0.95      0.96        36
weighted avg       0.97      0.97      0.97        36



/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:460: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


# 역시나 알 수가 없다...  
- 오류가 왜 발생하는지도 모르겠다.. 오류가 발생한다는 건 적절한 방법이 아니라는 것이 아닐까?   
- SGD 오류는 학습부족 혹은 클래스 불균형 문제로 인해 발생하는 것 같다
  -  '''
   Precision(정밀도)와 F1-score가 일부 클래스에 대해 정의되지 않았음을 나타냅니다. 이 문제는 주로 특정 클래스에 대해 모델이 예측을 하지 않았을 때 발생합니다. 즉, 예측된 클래스 중 일부 클래스가 전혀 없기 때문에 해당 클래스에 대한 정밀도와 F1-score를 계산할 수 없는 경우입니다.
  '''  

- 오류가 발생한 logisticregression에 대해 GPT에 물어보니 데이터가 스케일링이 되지 않아서 이런 오류가 뜬다고 함
  - ''' 이 오류 메시지는 Logistic Regression 모델이 주어진 데이터로 충분히 수렴하지 않았음을 나타냅니다. 일반적으로 이런 문제가 발생하는 이유는:

    데이터가 스케일링되지 않음:

      Logistic Regression은 데이터 스케일에 민감합니다. 만약 입력 데이터의 특성(feature) 값들이 서로 다른 스케일(예: 일부는 1에서 1000, 일부는 0에서 1)이라면, 모델이 최적의 솔루션을 찾기 어렵습니다.
      이를 해결하기 위해 데이터 스케일링(표준화 또는 정규화)을 적용해야 합니다. 예를 들어, StandardScaler 또는 MinMaxScaler를 사용해 데이터를 스케일링할 수 있습니다.
    최대 반복 횟수 부족:

      Logistic Regression 모델이 수렴하지 않은 이유 중 하나는 반복 횟수(max_iter)가 부족하기 때문일 수 있습니다. 기본적으로 max_iter=100인데, 이는 데이터에 따라 충분하지 않을 수 있습니다. 반복 횟수를 늘려서 문제를 해결할 수 있습니다
'''

# GPT의 답변을 첨부합니다.  
- Scikit-learn에서 제공하는 **Wine 데이터셋**(`load_wine`)은 다양한 머신러닝 모델을 테스트하고 비교할 수 있는 좋은 예제입니다. 와인의 화학 성분을 기반으로 세 가지 와인 품종을 분류하는 문제이며, 이 문제에 대해 어떤 예측 모델이 가장 효율적인지는 여러 요인에 따라 달라질 수 있습니다. 일반적으로는 모델의 **성능(정확도)**, **훈련 시간**, **해석 용이성**, **모델의 복잡도** 등을 고려해야 합니다.

다섯 가지 모델을 비교할 때 각각의 특성과 Wine 데이터셋에 적합한 이유를 간단히 설명하겠습니다.

### 1. **SGDClassifier (Stochastic Gradient Descent)**
   - **특징**: 확률적 경사 하강법을 사용하여 선형 분류기를 학습합니다. 데이터가 매우 큰 경우 효율적입니다.
   - **장점**: 대규모 데이터셋에서 학습이 매우 빠르며, 특히 희소 데이터에 적합합니다.
   - **단점**: Wine 데이터셋은 비교적 작기 때문에, SGDClassifier가 최적의 성능을 보이지 않을 수 있습니다.
   - **추천**: 작은 데이터셋에서는 비효율적일 수 있으므로 Wine 데이터셋에서는 최적의 선택이 아닐 수 있습니다.

### 2. **SVM (Support Vector Machine)**
   - **특징**: 데이터가 선형적으로 분리되지 않는 경우에도 커널을 사용해 고차원 공간에서 분류 가능.
   - **장점**: 작은 데이터셋에서는 매우 좋은 성능을 발휘할 수 있습니다. 특히 데이터가 선형적으로 분리되지 않을 때 뛰어난 성능을 발휘합니다.
   - **단점**: 데이터셋이 커지면 학습 속도가 느려지고, 메모리 사용량이 증가할 수 있습니다.
   - **추천**: Wine 데이터셋처럼 작은 크기에서는 좋은 성능을 발휘할 가능성이 큽니다.

### 3. **RandomForestClassifier**
   - **특징**: 여러 개의 결정 트리를 학습하여 앙상블 기법을 통해 예측하는 모델입니다.
   - **장점**: 매우 강력한 모델로, 과적합 방지에도 강하고, 데이터 전처리의 필요성이 적습니다.
   - **단점**: 매우 복잡한 모델로, 학습 시간이 길어질 수 있습니다.
   - **추천**: Wine 데이터셋에서 높은 정확도를 기대할 수 있으며, 대체로 성능이 매우 좋습니다. 앙상블 모델이므로 비교적 안정적이고 정확한 결과를 제공합니다.

### 4. **DecisionTreeClassifier**
   - **특징**: 데이터에 대한 의사결정 규칙을 기반으로 트리를 형성하여 분류를 수행합니다.
   - **장점**: 매우 직관적이고 해석하기 쉬우며, 학습 속도가 빠릅니다.
   - **단점**: 트리 깊이를 제한하지 않으면 과적합될 수 있습니다. 이는 작은 데이터셋에서도 문제로 작용할 수 있습니다.
   - **추천**: Wine 데이터셋에 빠르게 적용할 수 있으며, 파라미터를 잘 조정하면 괜찮은 성능을 기대할 수 있지만, 과적합 문제에 주의해야 합니다.

### 5. **LogisticRegression**
   - **특징**: 선형 모델로, 클래스 간의 경계를 직선(또는 초평면)으로 구분합니다.
   - **장점**: 간단하고 해석이 쉬운 모델입니다. 작은 데이터셋에서도 빠르게 수렴하며, 과적합을 방지하기 위한 정규화를 지원합니다.
   - **단점**: 데이터가 선형적으로 구분되지 않으면 성능이 떨어질 수 있습니다.
   - **추천**: Wine 데이터셋에서 빠르고 간단한 방법으로 학습할 수 있으며, 기본적인 성능은 충분히 보장합니다.

### 비교 분석

- **데이터 크기**: Wine 데이터셋은 비교적 작고, 약간의 비선형성 및 상관관계가 존재합니다. 이 때문에 매우 복잡한 모델(SGDClassifier)보다 적당한 복잡도의 모델이 더 적합할 수 있습니다.
- **모델 해석**: 해석이 용이한 모델을 원한다면 **Logistic Regression**이나 **Decision Tree**가 좋습니다.
- **최고의 성능**: **RandomForestClassifier**나 **SVM**이 가장 높은 성능을 발휘할 가능성이 큽니다. 특히 **RandomForest**는 과적합을 방지하면서 높은 성능을 내기 때문에 일반적으로 많은 데이터셋에서 좋은 결과를 제공합니다.
- **복잡도와 시간**: **Decision Tree**는 학습 시간이 짧고, 해석이 용이하다는 장점이 있지만, 과적합 가능성이 있습니다. **RandomForest**는 안정적이지만 상대적으로 시간이 조금 더 소요됩니다.

### 결론

Wine 데이터셋에서 **최고의 성능**을 기대하려면:
- **RandomForestClassifier** 또는 **SVM**을 사용하는 것이 가장 효율적일 가능성이 큽니다.

만약 성능과 함께 **모델 해석 용이성**도 고려해야 한다면:
- **Logistic Regression** 또는 **Decision Tree**가 적합합니다.

성능이 최우선이라면 **RandomForestClassifier**가 가장 추천됩니다.

# Project 3

In [62]:
# (1) 필요한 모듈 import
from sklearn.datasets import load_breast_cancer #싸이킷런 데이터셋에 있는 breast_cancer를 불러오는 코드
from sklearn.model_selection import train_test_split #싸이킷런에 있는 데이터를 train과 test dataset으로 나누어주는 함수 import
from sklearn.tree import DecisionTreeClassifier # 싸이킷런에 있는 결정트리분류기를 사용하기 위해 불러오는 코드
from sklearn.metrics import classification_report #싸이킷런에 있는 분류 결과에 대한 시각화를 위해 쓰는 코드

# (2) 데이터 준비
breast_cancer = load_breast_cancer() #breast_cancer 데이터 전체를 불러온다.
breast_cancer_data = breast_cancer.data #breast_cancer데이터의 data컬럼을 분류해 breast_cancer_data 변수에 담는다.
breast_cancer_label = breast_cancer.target #breast_cancer데이터의 target컬럼을 분류해 breast_cancer_label 변수에 담는다.

# (3) train, test 데이터 분리
#train_test_split()를 사용하여 X값, y값을 각각 train data와 test data로 나눈다. 함수에 들어 갈 파라미터로는 x,y가 들어가고
#test_size는 몇대몇으로 나눌지 정하는 옵션, random_state는 랜덤 패턴의 값을 지정한다. (어떤 값을 넣어도 무방하다.)
X_train, X_test, y_train, y_test = train_test_split(breast_cancer_data,
                                                    breast_cancer_label,
                                                    test_size=0.25,
                                                    random_state=2)


# (4) 모델 학습 및 예측
decision_tree = DecisionTreeClassifier(random_state=28) #결정트리분류기의 객체를 만든다.
decision_tree.fit(X_train, y_train) # 분류기에 x와 y의 훈련 데이터를 넣어 훈련 시킨다.
y_pred = decision_tree.predict(X_test) # 훈련된 분류기에 X_test라는 테스트 데이터셋을 넣어 얼마나 예측했는지 확인한다.

print(classification_report(y_test, y_pred)) # 결과를 지표로 확인하기 위해 classification_report를 활용해 y_test, y_pred 값을 넣어 확인한다.

              precision    recall  f1-score   support

           0       0.89      0.96      0.92        56
           1       0.98      0.92      0.95        87

    accuracy                           0.94       143
   macro avg       0.93      0.94      0.93       143
weighted avg       0.94      0.94      0.94       143



In [63]:
random_forest = RandomForestClassifier(random_state=2) # RandomForest분류기 객체를 생성
random_forest.fit(X_train, y_train) # 훈련
y_pred = random_forest.predict(X_test) # 예측

print(classification_report(y_test, y_pred)) # 결과 지표를 확인

              precision    recall  f1-score   support

           0       0.90      0.95      0.92        56
           1       0.96      0.93      0.95        87

    accuracy                           0.94       143
   macro avg       0.93      0.94      0.93       143
weighted avg       0.94      0.94      0.94       143



In [64]:
from sklearn import svm #Support Vector Machine을 사용하기 위해 import
svm_model = svm.SVC() # 모델 객체를 만든다.

print(svm_model._estimator_type) # 이 모델이 어떤 타입인지 확인한다.

svm_model.fit(X_train, y_train) # 훈련
y_pred = svm_model.predict(X_test) # 예측

print(classification_report(y_test, y_pred)) # 결과 지표를 확인


classifier
              precision    recall  f1-score   support

           0       0.92      0.86      0.89        56
           1       0.91      0.95      0.93        87

    accuracy                           0.92       143
   macro avg       0.92      0.91      0.91       143
weighted avg       0.92      0.92      0.92       143



In [65]:
from sklearn.linear_model import SGDClassifier #선형분류기인 SGDClassifier를 사용하기 위한 import
sgd_model = SGDClassifier() # 모델 객체 생성

print(sgd_model._estimator_type) # 이 모델의 타입을 확인
sgd_model.fit(X_train, y_train) # sgd모델로 훈련데이터로 훈련시킨다.
y_pred = sgd_model.predict(X_test)# 그 모델로 test데이터를 사용해 예측

print(classification_report(y_test, y_pred)) # 결과 지표를 확인

classifier
              precision    recall  f1-score   support

           0       0.76      0.93      0.84        56
           1       0.95      0.82      0.88        87

    accuracy                           0.86       143
   macro avg       0.86      0.87      0.86       143
weighted avg       0.88      0.86      0.86       143



In [66]:
from sklearn.linear_model import LogisticRegression # 선형분류기인 LogisticRegression를 사용하기 위한 import
logistic_model = LogisticRegression() # 모델 객체 생성

print(logistic_model._estimator_type) # 이 모델의 타입을 확인

logistic_model.fit(X_train, y_train) #LogisticRegression모델로 훈련데이터를 가지고 훈련시킨다.
y_pred = logistic_model.predict(X_test) # 예측

print(classification_report(y_test, y_pred)) # 결과 지표를 확인

classifier
              precision    recall  f1-score   support

           0       0.91      0.93      0.92        56
           1       0.95      0.94      0.95        87

    accuracy                           0.94       143
   macro avg       0.93      0.94      0.93       143
weighted avg       0.94      0.94      0.94       143



/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:460: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


# 흠  
- 오류가 발생하지 않은 나머지 것들 중 적용해야 맞지 않을까?  
- 머신러닝의 과정에 대한 이해가 없이 진행하니 사실 이해가 되지 않는다.
- 머신러닝의 기초에 관한 노드학습이 먼저 진행되었다면 좀 더 좋지 않았을 까 하는 아쉬움이 들지만, 아쉬움을 표현하기에는 매번 나는 이해를 잘 못하는 것 같다...


# GPT의 답변  
Scikit-learn에서 제공하는 **Breast Cancer 데이터셋**(`load_breast_cancer`)은 유방암 진단을 위한 데이터셋으로, 종양이 양성인지 악성인지를 분류하는 문제입니다. 이 데이터셋은 **2개의 클래스(양성, 악성)**로 구성된 **이진 분류 문제**이며, 총 **30개의 특징(feature)**을 가지고 있습니다. 데이터셋 크기도 적당해서 여러 모델을 테스트하기에 좋은 데이터셋입니다.

다섯 가지 모델을 비교할 때, Breast Cancer 데이터셋에 적합한 이유를 설명하겠습니다.

### 1. **SGDClassifier (Stochastic Gradient Descent)**
   - **특징**: 확률적 경사 하강법(SGD)을 사용하여 선형 분류기를 학습합니다.
   - **장점**: 매우 큰 데이터셋에서 학습이 빠르고, 특히 희소 데이터에 적합합니다.
   - **단점**: Breast Cancer 데이터셋은 비교적 작은 크기이기 때문에, SGD의 장점인 빠른 학습 속도가 필요하지 않을 수 있습니다. 작은 데이터셋에서는 다른 모델보다 성능이 떨어질 수 있습니다.
   - **추천**: Breast Cancer 데이터셋처럼 크기가 작고 비선형적인 관계가 있는 데이터에서는 SGD보다 다른 모델이 더 적합할 수 있습니다.

### 2. **SVM (Support Vector Machine)**
   - **특징**: SVM은 데이터가 선형적으로 분리되지 않는 경우에도 커널을 사용해 고차원 공간에서 분류할 수 있습니다.
   - **장점**: 작은 데이터셋에서 매우 높은 성능을 발휘할 수 있습니다. 특히 데이터가 선형적으로 완벽히 분리되지 않는 경우, SVM이 뛰어난 성능을 보일 수 있습니다.
   - **단점**: 데이터셋이 커질수록 학습 속도가 느려질 수 있지만, Breast Cancer 데이터셋은 크기가 작기 때문에 문제가 되지 않습니다.
   - **추천**: Breast Cancer 데이터셋에서는 **SVM**이 매우 높은 성능을 보일 가능성이 큽니다.

### 3. **RandomForestClassifier**
   - **특징**: 여러 개의 결정 트리를 학습하여 앙상블 기법을 통해 예측하는 모델입니다.
   - **장점**: 과적합을 방지하면서도 매우 강력한 성능을 제공합니다. 데이터 전처리가 거의 필요하지 않으며, 결측치에 덜 민감합니다.
   - **단점**: 트리의 수가 많아지면 학습 시간이 조금 늘어날 수 있지만, Breast Cancer 데이터셋은 크기가 작아서 학습 시간은 문제가 되지 않습니다.
   - **추천**: 이 데이터셋에서 **RandomForestClassifier**는 매우 좋은 성능을 발휘할 수 있습니다. 앙상블 모델이기 때문에 높은 정확도와 안정성을 기대할 수 있습니다.

### 4. **DecisionTreeClassifier**
   - **특징**: 의사결정 규칙을 기반으로 트리를 형성하여 분류를 수행합니다.
   - **장점**: 매우 직관적이고 해석하기 쉬우며, 학습 속도가 빠릅니다. 소규모 데이터셋에 적합하고, 모델의 구조가 해석하기 쉬워 의사결정의 근거를 설명하기 용이합니다.
   - **단점**: 과적합되기 쉽습니다. 특히 트리 깊이를 제한하지 않으면 성능이 떨어질 수 있습니다.
   - **추천**: Breast Cancer 데이터셋에서는 적당한 트리 깊이 제한을 설정하여 과적합을 방지하면 **빠르고 직관적인 결과**를 얻을 수 있습니다.

### 5. **LogisticRegression**
   - **특징**: 선형 분류 모델로, 클래스 간 경계를 선형으로 구분합니다.
   - **장점**: 간단하고 빠르게 학습할 수 있으며, 이진 분류 문제에 적합합니다. 정규화 기법(L2 정규화)을 통해 과적합을 방지할 수 있습니다.
   - **단점**: 데이터가 선형적으로 분리되지 않으면 성능이 떨어질 수 있습니다.
   - **추천**: Breast Cancer 데이터셋에서 **Logistic Regression**은 매우 좋은 선택입니다. 모델이 간단하지만 이진 분류 문제에서 강력한 성능을 발휘합니다.

### 비교 분석

- **데이터 크기**: Breast Cancer 데이터셋은 569개의 샘플과 30개의 특징을 가지고 있으며, 상대적으로 작은 데이터셋입니다. 모든 모델이 적절하게 작동할 수 있는 크기입니다.
- **모델 해석**: **Logistic Regression**과 **Decision Tree**는 해석이 쉬운 모델입니다. 특히 **Decision Tree**는 모델이 어떻게 의사결정을 하는지 직관적으로 파악할 수 있습니다.
- **최고의 성능**: **SVM**과 **RandomForestClassifier**가 가장 높은 성능을 발휘할 가능성이 큽니다. 특히 **RandomForestClassifier**는 데이터 전처리가 적고 안정적인 성능을 제공합니다.
- **복잡도와 시간**: **Decision Tree**는 학습 시간이 짧고 빠르며, **RandomForestClassifier**는 성능이 뛰어나지만 시간이 조금 더 걸릴 수 있습니다.

### 결론

Breast Cancer 데이터셋에서 **최고의 성능**을 기대하려면:
- **SVM** 또는 **RandomForestClassifier**가 가장 효율적일 가능성이 큽니다. 특히 **SVM**은 데이터가 작을 때 매우 강력한 성능을 발휘합니다.

모델 해석의 용이성도 중요하다면:
- **Logistic Regression**과 **Decision Tree**가 적합합니다.

성능을 최우선으로 한다면 **SVM** 또는 **RandomForestClassifier**가 **Breast Cancer 데이터셋**에서 높은 정확도를 기대할 수 있습니다. **Logistic Regression**은 이진 분류 문제에 적합하고 빠르게 학습할 수 있어 실용적입니다.